model/dataProvider/SyncDataProviderCollection.php

```diff
@@ -0,0 +1,166 @@
+<?php
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019  (original work) Open Assessment Technologies SA;
+ *
+ * @author Yuri Filippovich
+ */
+
+namespace oat\taoSync\model\dataProvider;
+
+use oat\oatbox\service\ConfigurableService;
+use oat\taoSync\model\Exception\SyncDataProviderException;
+use Zend\ServiceManager\ServiceLocatorInterface;
+
+class SyncDataProviderCollection extends ConfigurableService
+{
+    const SERVICE_ID = 'taoSync/SyncDataProviderCollection';
+
+    const OPTION_DATA_PROVIDERS = 'taoSync/dataProviders';
+
+    private $dataProviders;
+
+    private $rootDataProviders = [];
+    private $childDataProviders = [];
+
+    /**
+     * @inheritDoc
+     */
+    public function setServiceLocator(ServiceLocatorInterface $serviceLocator = null)
+    {
+        parent::setServiceLocator($serviceLocator);
+        $this->initDataProviders();
```

Comment:
I don't really get this logic with initialization of data providers and building a tree. But if we really need it I think it would be better to initialize it when we try to use it for the first time. `setServiceLocator` should be only for setting ServiceLocator.

Reply:
changed